File: python_mocra/engine/task/task.py
from typing import List, Optional, Any, Dict
from uuid import UUID
import asyncio
from pydantic import BaseModel

from common.models.entity import AccountModel, PlatformModel
from common.models.message import Request
from engine.task.module import Module

class Task:
    """
    Enhanced Task Structure.
    Contains complete configuration, modules, and execution context.
    """
    def __init__(self,
                 account: AccountModel,
                 platform: PlatformModel,
                 modules: List[Module],
                 metadata: Dict[str, Any],
                 run_id: UUID,
                 prefix_request: UUID,
                 login_info: Any = None):
        self.account = account
        self.platform = platform
        self.modules = modules
        self.metadata = metadata
        self.run_id = run_id
        self.prefix_request = prefix_request
        self.login_info = login_info

    @property
    def id(self) -> str:
        return f"{self.account.name}-{self.platform.name}"

    def get_module_names(self) -> List[str]:
        return [m.id for m in self.modules]
    
    def is_empty(self) -> bool:
        return len(self.modules) == 0

    async def build_requests(self, original_task_model: Any) -> List[Request]:
        """
        Concurrently execute generation for all modules.
        """
        all_requests = []
        
        # In Python we don't have join_all generally available for generators easily unless we wrap them.
        # We iterate sequentially or use asyncio.gather on wrapped coroutines.
        # Since module.generate is an AsyncGenerator, we consume it.
        
        tasks = []
        for module in self.modules:
            tasks.append(self._collect_module_requests(module, original_task_model))
            
        results = await asyncio.gather(*tasks, return_exceptions=True)
        
        for res in results:
            if isinstance(res, list):
                all_requests.extend(res)
            elif isinstance(res, Exception):
                # Log error but don't crash everything? 
                # Rust returns Err(e) to caller.
                raise res
                
        return all_requests

    async def _collect_module_requests(self, module: Module, task_model: Any) -> List[Request]:
        reqs = []
        async for req in module.generate(task_model):
            reqs.append(req)
        return reqs

File: engine/src/task/task.rs
#![allow(unused)]
use common::model::entity::{AccountModel, PlatformModel};
use common::model::{Request, login_info::LoginInfo};
use crate::task::module::Module;
use errors::Result;
use uuid::Uuid;
use futures::StreamExt;

/// 增强的任务结构，包含更完整的配置信息
#[derive(Clone)]
pub struct Task {
    pub account: AccountModel,
    pub platform: PlatformModel,
    pub login_info: Option<LoginInfo>,
    pub modules: Vec<Module>,
    pub metadata: serde_json::Map<String, serde_json::Value>,
    pub run_id:Uuid,
    pub prefix_request:Uuid
}


impl Task {
    /// 获取任务的唯一标识符
    pub fn id(&self) -> String {
        format!("{}-{}", self.account.name, self.platform.name)
    }

    /// 获取工作模块的名称列表
    pub fn get_module_names(&self) -> Vec<String> {
        self.modules.iter().map(|module| module.id()).collect()
    }

    /// 并发执行所有模块的请求生成
    pub async fn build_requests(&self) -> Result<Vec<Request>> {
        use futures::future::join_all;
        // 1) 并发创建每个模块的请求流
        let futures = self
            .modules
            .iter()
            .map(|module| module.generate(self.metadata.clone(), self.login_info.clone()));

        let results = join_all(futures).await;

        // 2) 合并所有请求流
        let mut all_requests = Vec::new();
        for result in results {
            match result {
                Ok(stream) => {
                    let mut reqs: Vec<Request> = stream.collect().await;
                    all_requests.append(&mut reqs);
                }
                Err(e) => return Err(e),
            }
        }
        Ok(all_requests)
    }
    pub fn is_empty(&self)->bool{
        self.modules.is_empty()
    }
}
